doc/design/to_run.md

```diff
@@ -0,0 +1,165 @@
+a table/view.  However, some pre-processing transformations are not row-wise.
+An example is *sliding-window*.
+
+A [user told us](https://github.com/sql-machine-learning/sqlflow/issues/2238)
+that he wants to extract temporal features from time-series data by calling the
+well-known Python package [`tsfresh`](https://tsfresh.readthedocs.io), which
+runs a sliding-window over the time-series data and convert each step into a new
+row.
+
+This requirement intrigues us to support a general way to call arbitrary
+programs from SQL, hence the idea of the `TO RUN` clause.
+
+
+## Syntax Extension
+
+The subject to run is a program.  A program needs versioning and releasing.  As
+always, we assume Docker images are released form.  Therefore, the user needs to
+provide a Docker image as the subject of the `TO RUN` clause.  The SQLFlow
+compiler translates the TO RUN statement into a Tekton step that runs this
+Docker image.
+
+```sql
+SELECT * FROM source_table ORDER BY creation_date
+TO RUN 
+  a_data_scientist/extract_ts_features:1.0
+CMD 
+  "--verbose",
+  "--window_width=120"
+  "output_table_name";
+```
+
+In this example, the entrypoint program is likely written in Python so it can
+call the `tsfresh` Python package.  The CMD clause provides command-line options
+to the entrypoint program.
+
+
+### The SELECT Prefix
+
+As the `TO RUN` clause can run any program in the Docker image, it is not
+necessary to have the SELECT statement as a prefix.  Instead, we can change the
+above syntax design to make the SELECT prefix a command-line option to the
+entrypoint program, which then calls the DBMS API to execute it.
+
+```sql
+TO RUN 
+  a_data_scientist/extract_ts_features:1.0
+CMD 
+  "SELECT * FROM source_table ORDER BY creation_date"
+  "--verbose",
+  "--window_width=120"
+  "output_table_name";
+```
+
+However, we still prefer the SELECT statement as a prefix, but the SQLFlow
+compiler doesn't run it as a step container; instead, the compiler passes the
+SELECT statement to the entrypoint as part of the context.
+
+
+## The Context
+
+In the above example, the entrypoint program takes three command-line options:
+`--verbose`, `--window_width=120`, and `output_table_name`.  Also, the program
+needs context information, including the DBMS endpoints, credential information
+to access the data, and the SELECT prefix.
+
+The SQLFlow compiler has to pass context in the form of environment variables
+other than command-line options because some command-line parsing frameworks
+terminates the program seeing unknown options.  When SQLFlow upgrades and
+introduces new options, the entrypoint program would fail.
+
+The SQLFlow server cannot pass in all context information in a single
+environment variable, which has a limit of value size.  Instead, it sets
+environment variables prefixed with `SQLFLOW_`.
+
+- `SQLFLOW_DB`: the type of DBMS.
+- [To-be-complete]
+
+
+## Run a Python Function
```

Comment:
The user will provide a python program instead of python function.
The python program contains a `main` function. The main function will parse the command line parameters and execute the table I/O, data computing process.

Reply:
It a Python function, but not a Python program, that the user should provide. The entrypoint programs I added to this section can `eval` the function call expression.  This example entrypoint program cannot handle table I/O and data computing. But this example shows that users, if they want, could provide more powerful entrypoint programs to handle anything, including starting Dask/Mars jobs.